### PR TITLE
feat(examples): add Next.js example

### PR DESCRIPTION
**Pull request by duarten**
Adds a small Javascript-only example for Next.js, based on the official "Getting Started" guide.

Refs #2554

Note that this PR requires #2525 to be merged.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [X] Other... Please describe: Example.

## What is the current behavior?

Issue Number: #2554

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by flolu** [comment]
I've just tried it, but it fails with this error:
```
  10 |
  11 |       <main>
> 12 |         <h1 className="title">
     |                             ^
  13 |           Welcome to <a href="https://nextjs.org'>Next.js!</a'
  14 |         </h1>
  15 |


> Build error occurred
Error: > Build failed because of webpack errors
    at build (home/flo/.cache/bazel/_bazel_flo/c803801ec4119bfe034ba71f43e5fbde/sandbox/linux-sandbox/8/execroot/examples_nextjs/node_modules/next/build/index.ts:452:11)
```

**Comment by flolu** [comment]
Also, what do you think of [this](https://github.com/flolu/nextjs-bazel) implementation?

Everything works fine, except for global styles ([the issue](https://github.com/flolu/nextjs-bazel/issues/1))
Maybe we can use your `expand_template` to fix it?

**Comment by duarten** [comment]
> I've just tried it, but it fails with this error:
> 
> ```
>   10 |
>   11 |       <main>
> > 12 |         <h1 className="title">
>      |                             ^
>   13 |           Welcome to <a href="https://nextjs.org'>Next.js!</a'
>   14 |         </h1>
>   15 |
> 
> 
> > Build error occurred
> Error: > Build failed because of webpack errors
>     at build (home/flo/.cache/bazel/_bazel_flo/c803801ec4119bfe034ba71f43e5fbde/sandbox/linux-sandbox/8/execroot/examples_nextjs/node_modules/next/build/index.ts:452:11)
> ```

Strange, can't image what the issue is. Works fine for me with the most up-do-date rules_nodejs and with the fix in #2525 applied.

**Comment by duarten** [comment]
> Also, what do you think of [this](https://github.com/flolu/nextjs-bazel) implementation?
> 
> Everything works fine, except for global styles ([the issue](https://github.com/flolu/nextjs-bazel/issues/1))
> Maybe we can use your `expand_template` to fix it?

It looks good in general. I had to work around other issues though. For example, if you start importing react modules, webpack will bundle multiple react versions unless `--bazel_run_from_execroot` is specified (and symlinks are preserved). That took me a while to figure out.

**Comment by flolu** [comment]
> Strange, can't image what the issue is. Works fine for me with the most up-do-date rules_nodejs and with the fix in #2525 applied.

Oh, I probably forgot to apply #2525. How can I patch rules_nodejs?

File: examples/nextjs/pages/index.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
+import Head from 'next/head'
+
+export default function Home() {
+  return (
+    <div className='container'>
+      <Head>
+        <title>Create Next App</title>
+        <link rel="icon" href="/favicon.ico' /'
```
Comment (flolu) [inline examples/nextjs/pages/index.js:8]: Ahh, here is the problem. This line should be

```
<link rel="icon" href="/favicon.ico">
```

File: examples/nextjs/pages/index.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+      </Head>
+
+      <main>
+        <h1 className="title">
+          Welcome to <a href="https://nextjs.org'>Next.js!</a'
```
Comment (flolu) [inline examples/nextjs/pages/index.js:13]: Also should be
```
Welcome to <a href="https://nextjs.org'>Next.js!"></a>
```

**Review by flolu — COMMENTED** [review]
There were just some issues in the html template.
After fixing them, it works fine.

**But** after adding global styles I am encountering the same issue as here: flolu/nextjs-bazel#1
```
./styles/globals.css
Global CSS cannot be imported from files other than your Custom <App>. Please move all global CSS imports to pages/_app.js. Or convert the import to Component-Level CSS (CSS Modules).
Read more: https://nextjs.org/docs/messages/css-global
Location: pages/_app.jsx
```


The problem is that [this if statement](https://github.com/vercel/next.js/blob/cd9bb986eba25006d4872e30eedf24ef5680bc6b/packages/next/server/lib/find-page-file.ts#L34) fails with `Error: EACCES: permission denied.`

Do you have an idea to make this work?

File: examples/nextjs/pages/index.js (hunk 1 above)
Comment (duarten) [reply examples/nextjs/pages/index.js:8]: Yes, it was the format step; this is from the pre-commit hook:

```
-          href="https://vercel.com?utm_source=create-next-app&utm_medium=default-template&utm_campaign=create-next-app"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Powered by{' '}
-          <img src="/vercel.svg" alt="Vercel Logo" className="logo" />
-        </a>
+          href="https://vercel.com?utm_source=create-next-app&utm_medium=default-template&utm_campaign=create-next-app'
+          target='_blank'
+  rel = 'noopener noreferrer' >
+      Powered by{' '}<img src = '/vercel.svg' alt = 'Vercel Logo' className = 'logo' />
+      </a>
       </footer>

       <style jsx>{`
@@ -203,7 +201,7 @@ export default function Home() {
         * {
           box-sizing: border-box;
         }
-      `}</style>
-    </div>
+      `}</style><
+      /div>
   )
 }

      ERROR: please run yarn format to format changes in your commit
```

**Comment by flolu** [comment]
Do you have any idea regarding the global styles issue? https://github.com/bazelbuild/rules_nodejs/pull/2577#pullrequestreview-625028842

**Comment by duarten** [comment]
> Do you have any idea regarding the global styles issue? [#2577 (review)](https://github.com/bazelbuild/rules_nodejs/pull/2577#pullrequestreview-625028842)

Interestingly, it works fine for me with `bzl run //example.dev`, but when I do `bzl build //example`, it fails with:

```
Failed to compile.

./style/global.css
Global CSS cannot be imported from files other than your Custom <App>. Please move all global CSS imports to pages/_app.js. Or convert the import to Component-Level CSS (CSS Modules).
Read more: https://nextjs.org/docs/messages/css-global
Location: pages/_app.js
```

Notice how the location is set to "pages/_app.js".

These are my changes:

`style/global.css`:
```css
html,
body {
  background: black;
}
```

`pages/_app.js`:
```js
import '../style/global.css'

export default function MyApp({ Component, pageProps }) {
  return <Component {...pageProps} />
}
```

```diff
diff --git a/examples/nextjs/index.bzl b/examples/nextjs/index.bzl
index 33df5645..e6e67cef 100644
--- a/examples/nextjs/index.bzl
+++ b/examples/nextjs/index.bzl
@@ -24,7 +24,7 @@ def nextjs(name, deps = [], visibility = ["//visibility:public"]):
         visibility: The visibility of the resulting targets (next.js output dir and public files).
     """
     inputs = "%s-srcs" % name
-    native.filegroup(name = inputs, srcs = native.glob(["pages/**"]))
+    native.filegroup(name = inputs, srcs = native.glob(["pages/**", "style/**"]))
 
     native.filegroup(
         name = "public_files",
```

**Comment by flolu** [comment]
Yeah, same for me. I've debugged this problem and the root cause it [this if statement](https://github.com/vercel/next.js/blob/cd9bb986eba25006d4872e30eedf24ef5680bc6b/packages/next/server/lib/find-page-file.ts#L34). It checks if the file can be accessed. But it fails with an `Error: EACCES: permission denied`.

Patching Nest.js to ignore the error seems to work. But who knows which side effects this might have.

Do you know why it throws `Error: EACCES: permission denied`?

**Comment by duarten** [comment]
I would to investigate this more to have a proper answer, but since it works with `bazel run` and not `bazel build`, and the way the files are laid out is not the same for both, it's probably related to that. In one case the symlink doesn't appear to have write permissions.

**Pull request closed without merging**